#7 (RFC) AJAX Flash Header feature only if HTTP request contains "X-Get-Flash: yes"

Ali1 opened the pull request:
#7

Ali1 committed:
AJAX Flash Header feature only if HTTP request contains "X-Get-Header: yes"

Ali1 commented:
Something like this. I don't see why anyone would need to customise the Header Key other than the previous need to make it empty - which is no longer necessary.

dereuromark commented:
Might need a small test adjustment, but looks good.

Ali1 commented:
Couldn't figure out to add request headers to testAjax

Ali1 committed:
fix tests

Ali1 commented:
done

dereuromark commented on composer.json:
```diff
@@ -17,7 +17,8 @@
 	"require-dev": {
-		"fig-r/psr2r-sniffer": "dev-master"
+		"fig-r/psr2r-sniffer": "dev-master",
+		"phpunit/phpunit": "~8.5.0"
```
we dont need this if we use the documented composer commands.

Ali1 replied:
not required for travis, but it's required to develop the code in this repo, no? is it usual practice to keep this out?

dereuromark replied:
See composer scripts section.

Ali1 committed:
Update README.md

Ali1 commented:
You reckon this feature would be useful for the core Flash Component? https://github.com/cakephp/cakephp

dereuromark commented:
I think so, it sounds quite useful to have the header invoke this.

Ali1 commented:
https://github.com/cakephp/cakephp/pull/14287

dereuromark commented on docs/README.md:
```diff
@@ -74,14 +74,25 @@ The following would only render (and remove) the error messages:
 
+### Use in AJAX requests
+
+Set the request header `X-Get-Flash: yes` in your AJAX calls to receive a json encoded response in the `X-Flash` with an array of keys containing an array of messages (each flash message contains a message key, a type key and params key):
```
instead of yes: Should we use
- `all` (both session and transient), we could also keep `yes` (just slightly less clear then)
- `transient`
- `session`

as options?
Usually only "transient" make really sense for AJAX, as those are generated by guarantee only in this request (no side effects). We wouldnt want other requests session based ones to get in here by accident anyway.

dereuromark replied:
Also, if transient is the default, then you dont need to manually provide the header, as you only get the ones of this request, no side effects possible.
That one you only need if you want to modify the default to include or use session. :) So everyone is happy.

Ali1 replied:
will consuming the transient messages in the component take them away from ajax view flash rendering? we wouldn't want the default to stop the messages reaching the view

Ali1 replied:
> Usually only "transient" make really sense for AJAX

Sometimes you use non-transient messages within controller methods that can be called ajax and normal. But this is not a bad options.... we can skip this whole request header in favour of recommending the use of transient messages

dereuromark replied:
Either way works for me.

Ali1 committed:
work on x-get-flash all/transient/session

Ali1 committed:
AJAX transient messages only

Ali1 committed:
coding

dereuromark commented on src/Controller/Component/FlashComponent.php:
```diff
@@ -55,29 +54,24 @@ public function beforeRender(Event $event) {
-					'params' => $message['params'],
-				];
-			}
+		foreach ($ajaxMessages as $message) {
```
Are we losing the key based structure here?

Ali1 committed:
add transient message easy calls

Ali1 commented:
So I've added helper addTransientSuccess/Info/Error and component transientSuccess/Info/Error methods via __call to make transient flash messages easier to use and I've removed the subarray structure of X-Flash output - because why should the client care about which internal key the controller uses to make a flash message and then the component uses to output the flash message - whichever key is used during that request will be the key that's relevant to the client side.

What do you think?

Ali1 committed:
Update FlashHelper.php

dereuromark commented on src/Controller/Component/FlashComponent.php:
```diff
@@ -20,14 +23,21 @@
  * @method void warning(string $message, array $options = []) Set a message using "warning" element
  * @method void info(string $message, array $options = []) Set a message using "info" element
+ * @method void transientSuccess(string $message, array $options = []) Set a message using "success" element.
+ *                                                                      These flash messages that are not saved (only available for current view)
```
saved => persisted across requests

You can also move this to the front and have 3 whitespace indentation.

Ali1 committed:
last commit

dereuromark commented:
I will have to test with Ajax plugin and https://github.com/dereuromark/cakephp-ajax/blob/master/src/Controller/Component/AjaxComponent.php#L44
And how they play together etc here.

Maybe Ajax also needs a fix on the transient part or could reuse Flash plugin here.

dereuromark commented:
Did you add unit tests for the magic method calls?
https://codecov.io/gh/dereuromark/cakephp-flash/pull/8/changes
Coverage drops like 5-6% here.

dereuromark commented:
Anything left here?
So we can merge and release 2.0 stable?

dereuromark commented:
I will merge to cake4 branch, lets do any follow up changes in that one.
And once this is all working, we can merge that to master and release/tag.

dereuromark commented:
And can we maybe add some show cases / demos in https://github.com/dereuromark/cakephp-sandbox/compare/flash ?
That would also be nice for documentation of e.g. AJAX handling.
